`dprint` fails due to `dprint-plugin-exec` error: `failed to fill whole buffer`

Version: 0.40.0

Today, I noticed my [GitHub Action failed](https://github.com/google/comprehensive-rust/actions/runs/5708687281/job/15483185826) with

```
Compiling https://plugins.dprint.dev/markdown-0.15.2.wasm
Compiling https://plugins.dprint.dev/toml-0.5.4.wasm
Compiling https://plugins.dprint.dev/json-0.17.2.wasm
[dprint-plugin-exec]
Error: Error reading message from stdin.

Caused by:
    failed to fill whole buffer
Error: Error reading message from stdin.

Caused by:
    failed to fill whole buffer
Error resolving plugin https://plugins.dprint.dev/exec-0.3.5.json: Failed plugin schema verification. This may indicate you are using an old version of the dprint CLI or plugin and should upgrade.: The plugin schema version was 4, but expected 5.
Error: Process completed with exit code 12
```

Based on the "The plugin schema version was 4, but expected 5", I figured I could fix this by updating the version of the plugin in the config file.

However, `dprint config update` fails with the same error when I run it locally (with `dprint` version 0.40.0).

I guess I have two questions:

- How can I update the plugin by hand?
- Can `dprint` learn to not error out on the existing config?

Thanks for any help!

--- a/crates/core/src/plugins/process/communicator.rs
+++ b/crates/core/src/plugins/process/communicator.rs
@@ -130,12 +130,30 @@ impl ProcessPluginCommunicator {
     let mut stdout_reader = MessageReader::new(child.stdout.take().unwrap());
     let mut stdin_writer = MessageWriter::new(child.stdin.take().unwrap());
 
-    let (mut stdout_reader, stdin_writer) = tokio::task::spawn_blocking(move || {
-      verify_plugin_schema_version(&mut stdout_reader, &mut stdin_writer)
-        .context("Failed plugin schema verification. This may indicate you are using an old version of the dprint CLI or plugin and should upgrade.")?;
-      Ok::<_, anyhow::Error>((stdout_reader, stdin_writer))
+    let (mut stdout_reader, stdin_writer, schema_version) = tokio::task::spawn_blocking(move || {
+      let schema_version = get_plugin_schema_version(&mut stdout_reader, &mut stdin_writer)
+        .context("Failed plugin schema verification. This may indicate you are using an old version of the dprint CLI or plugin and should upgrade")?;
+      Ok::<_, anyhow::Error>((stdout_reader, stdin_writer, schema_version))
     })
     .await??;
+
+    if schema_version != PLUGIN_SCHEMA_VERSION {
+      // kill the child to prevent it from ouputting to stderr
+      let _ = child.kill();
+      if schema_version < PLUGIN_SCHEMA_VERSION {
+        bail!(
+          "The plugin schema version was {}, but expected {}.\n\nUpgrade instructions: https://dprint.dev/plugins/upgrade-instructions",
+          schema_version,
+          PLUGIN_SCHEMA_VERSION
+        );
+      } else {
+        bail!(
+          "The plugin schema version was {}, but expected {}. Try running: dprint config update",
+          schema_version,
+          PLUGIN_SCHEMA_VERSION
+        );
+      }
+    }
 
     let stdin_writer = SingleThreadMessageWriter::for_stdin(stdin_writer);
     let context = Rc::new(Context {
@@ -386,10 +404,7 @@ impl ProcessPluginCommunicator {
   }
 }
 
-fn verify_plugin_schema_version<TRead: Read + Unpin, TWrite: Write + Unpin>(
-  reader: &mut MessageReader<TRead>,
-  writer: &mut MessageWriter<TWrite>,
-) -> Result<()> {
+fn get_plugin_schema_version<TRead: Read + Unpin, TWrite: Write + Unpin>(reader: &mut MessageReader<TRead>, writer: &mut MessageWriter<TWrite>) -> Result<u32> {
   // since this is the setup, use a lot of contexts to find exactly where it failed
   writer.send_u32(0).context("Failed asking for schema version.")?; // ask for schema version
   writer.flush().context("Failed flushing schema version request.")?;
@@ -397,16 +412,7 @@ fn verify_plugin_schema_version<TRead: Read + Unpin, TWrite: Write + Unpin>(
   if acknowledgement_response != 0 {
     bail!("Plugin response was unexpected ({acknowledgement_response}).");
   }
-  let plugin_schema_version = reader.read_u32().context("Could not read schema version.")?;
-  if plugin_schema_version != PLUGIN_SCHEMA_VERSION {
-    bail!(
-      "The plugin schema version was {}, but expected {}.",
-      plugin_schema_version,
-      PLUGIN_SCHEMA_VERSION
-    );
-  }
-
-  Ok(())
+  reader.read_u32().context("Could not read schema version.")
 }
 
 fn std_err_redirect(shutdown_flag: Arc<AtomicFlag>, stderr: ChildStderr, on_std_err: impl Fn(String) + Send + Sync + 'static) {

--- a/crates/core/src/plugins/process/communicator.rs
+++ b/crates/core/src/plugins/process/communicator.rs
@@ -142,13 +142,13 @@ impl ProcessPluginCommunicator {
       let _ = child.kill();
       if schema_version < PLUGIN_SCHEMA_VERSION {
         bail!(
-          "The plugin schema version was {}, but expected {}.\n\nUpgrade instructions: https://dprint.dev/plugins/upgrade-instructions",
+          "This plugin is too old to run in the dprint CLI (version was {}, but expected {}).\n\nUpgrade instructions: https://github.com/dprint/dprint/issues/731",
           schema_version,
           PLUGIN_SCHEMA_VERSION
         );
       } else {
         bail!(
-          "The plugin schema version was {}, but expected {}. Try running: dprint config update",
+          "Your dprint CLI is too old to run this plugin (version was {}, but expected {}). Try running: dprint config update",
           schema_version,
           PLUGIN_SCHEMA_VERSION
         );

--- a/crates/core/src/plugins/process/communicator.rs
+++ b/crates/core/src/plugins/process/communicator.rs
@@ -142,7 +142,7 @@ impl ProcessPluginCommunicator {
       let _ = child.kill();
       if schema_version < PLUGIN_SCHEMA_VERSION {
         bail!(
-          "This plugin is too old to run in the dprint CLI (version was {}, but expected {}).\n\nUpgrade instructions: https://github.com/dprint/dprint/issues/731",
+          "This plugin is too old to run in the dprint CLI and you will need to manually upgrade it (version was {}, but expected {}).\n\nUpgrade instructions: https://github.com/dprint/dprint/issues/731",
           schema_version,
           PLUGIN_SCHEMA_VERSION
         );

--- a/crates/core/src/plugins/process/message_processor.rs
+++ b/crates/core/src/plugins/process/message_processor.rs
@@ -376,8 +376,9 @@ fn schema_establishment_phase<TRead: Read + Unpin, TWrite: Write + Unpin>(stdin:
     bail!("Expected a schema version request of `0`.");
   }
 
-  // 2. The client responds with `0` (4 bytes) for success, then `4` (4 bytes) for the schema version.
+  // 2. The client responds with `0` (4 bytes) for success
   stdout.send_u32(0)?;
+  // 3. Then 4 bytes for the schema version
   stdout.send_u32(PLUGIN_SCHEMA_VERSION)?;
   stdout.flush()?;
 

--- a/dprint.json
+++ b/dprint.json
@@ -21,6 +21,6 @@
     "https://plugins.dprint.dev/markdown-0.15.3.wasm",
     "https://plugins.dprint.dev/toml-0.5.4.wasm",
     "https://plugins.dprint.dev/prettier-0.27.0.json@3557a62b4507c55a47d8cde0683195b14d13c41dda66d0f0b0e111aed107e2fe",
-    "https://plugins.dprint.dev/exec-0.4.2.json@ca3c3d1cc1105b4ac1c31afdb26791a90209f13a5950f8de00dcab5e5ee293a7"
+    "https://plugins.dprint.dev/exec-0.4.3.json@42343548b8022c99b1d750be6b894fe6b6c7ee25f72ae9f9082226dd2e515072"
   ]
 }

--- a/dprint.json
+++ b/dprint.json
@@ -21,6 +21,6 @@
     "https://plugins.dprint.dev/markdown-0.15.3.wasm",
     "https://plugins.dprint.dev/toml-0.5.4.wasm",
     "https://plugins.dprint.dev/prettier-0.27.0.json@3557a62b4507c55a47d8cde0683195b14d13c41dda66d0f0b0e111aed107e2fe",
-    "https://plugins.dprint.dev/exec-0.4.3.json@42343548b8022c99b1d750be6b894fe6b6c7ee25f72ae9f9082226dd2e515072"
+    "https://plugins.dprint.dev/exec-0.3.5.json@d687dda57be0fe9a0088ccdaefa5147649ff24127d8b3ea227536c68ee7abeab"
   ]
 }

--- a/dprint.json
+++ b/dprint.json
@@ -21,6 +21,6 @@
     "https://plugins.dprint.dev/markdown-0.15.3.wasm",
     "https://plugins.dprint.dev/toml-0.5.4.wasm",
     "https://plugins.dprint.dev/prettier-0.27.0.json@3557a62b4507c55a47d8cde0683195b14d13c41dda66d0f0b0e111aed107e2fe",
-    "https://plugins.dprint.dev/exec-0.3.5.json@d687dda57be0fe9a0088ccdaefa5147649ff24127d8b3ea227536c68ee7abeab"
+    "https://plugins.dprint.dev/exec-0.4.3.json@42343548b8022c99b1d750be6b894fe6b6c7ee25f72ae9f9082226dd2e515072"
   ]
 }

